// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: flyteidl/admin/common.proto

package admin

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"strings"
	"time"
	"unicode/utf8"

	"github.com/golang/protobuf/ptypes"

	core "github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/core"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = ptypes.DynamicAny{}

	_ = core.ResourceType(0)

	_ = core.ResourceType(0)

	_ = core.ResourceType(0)

	_ = core.ResourceType(0)
)

// define the regex for a UUID once up-front
var _common_uuidPattern = regexp.MustCompile("^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$")

// Validate checks the field values on NamedEntityIdentifier with the rules
// defined in the proto definition for this message. If any rules are
// violated, an error is returned.
func (m *NamedEntityIdentifier) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for Project

	// no validation rules for Domain

	// no validation rules for Name

	return nil
}

// NamedEntityIdentifierValidationError is the validation error returned by
// NamedEntityIdentifier.Validate if the designated constraints aren't met.
type NamedEntityIdentifierValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NamedEntityIdentifierValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NamedEntityIdentifierValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NamedEntityIdentifierValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NamedEntityIdentifierValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NamedEntityIdentifierValidationError) ErrorName() string {
	return "NamedEntityIdentifierValidationError"
}

// Error satisfies the builtin error interface
func (e NamedEntityIdentifierValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNamedEntityIdentifier.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NamedEntityIdentifierValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NamedEntityIdentifierValidationError{}

// Validate checks the field values on NamedEntityMetadata with the rules
// defined in the proto definition for this message. If any rules are
// violated, an error is returned.
func (m *NamedEntityMetadata) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for Description

	// no validation rules for State

	return nil
}

// NamedEntityMetadataValidationError is the validation error returned by
// NamedEntityMetadata.Validate if the designated constraints aren't met.
type NamedEntityMetadataValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NamedEntityMetadataValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NamedEntityMetadataValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NamedEntityMetadataValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NamedEntityMetadataValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NamedEntityMetadataValidationError) ErrorName() string {
	return "NamedEntityMetadataValidationError"
}

// Error satisfies the builtin error interface
func (e NamedEntityMetadataValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNamedEntityMetadata.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NamedEntityMetadataValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NamedEntityMetadataValidationError{}

// Validate checks the field values on NamedEntity with the rules defined in
// the proto definition for this message. If any rules are violated, an error
// is returned.
func (m *NamedEntity) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for ResourceType

	if v, ok := interface{}(m.GetId()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return NamedEntityValidationError{
				field:  "Id",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if v, ok := interface{}(m.GetMetadata()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return NamedEntityValidationError{
				field:  "Metadata",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	return nil
}

// NamedEntityValidationError is the validation error returned by
// NamedEntity.Validate if the designated constraints aren't met.
type NamedEntityValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NamedEntityValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NamedEntityValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NamedEntityValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NamedEntityValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NamedEntityValidationError) ErrorName() string { return "NamedEntityValidationError" }

// Error satisfies the builtin error interface
func (e NamedEntityValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNamedEntity.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NamedEntityValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NamedEntityValidationError{}

// Validate checks the field values on Sort with the rules defined in the proto
// definition for this message. If any rules are violated, an error is returned.
func (m *Sort) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for Key

	// no validation rules for Direction

	return nil
}

// SortValidationError is the validation error returned by Sort.Validate if the
// designated constraints aren't met.
type SortValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SortValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SortValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SortValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SortValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SortValidationError) ErrorName() string { return "SortValidationError" }

// Error satisfies the builtin error interface
func (e SortValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSort.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SortValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SortValidationError{}

// Validate checks the field values on NamedEntityIdentifierListRequest with
// the rules defined in the proto definition for this message. If any rules
// are violated, an error is returned.
func (m *NamedEntityIdentifierListRequest) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for Project

	// no validation rules for Domain

	// no validation rules for Limit

	// no validation rules for Token

	if v, ok := interface{}(m.GetSortBy()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return NamedEntityIdentifierListRequestValidationError{
				field:  "SortBy",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for Filters

	return nil
}

// NamedEntityIdentifierListRequestValidationError is the validation error
// returned by NamedEntityIdentifierListRequest.Validate if the designated
// constraints aren't met.
type NamedEntityIdentifierListRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NamedEntityIdentifierListRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NamedEntityIdentifierListRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NamedEntityIdentifierListRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NamedEntityIdentifierListRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NamedEntityIdentifierListRequestValidationError) ErrorName() string {
	return "NamedEntityIdentifierListRequestValidationError"
}

// Error satisfies the builtin error interface
func (e NamedEntityIdentifierListRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNamedEntityIdentifierListRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NamedEntityIdentifierListRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NamedEntityIdentifierListRequestValidationError{}

// Validate checks the field values on NamedEntityListRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, an error is returned.
func (m *NamedEntityListRequest) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for ResourceType

	// no validation rules for Project

	// no validation rules for Domain

	// no validation rules for Limit

	// no validation rules for Token

	if v, ok := interface{}(m.GetSortBy()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return NamedEntityListRequestValidationError{
				field:  "SortBy",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for Filters

	return nil
}

// NamedEntityListRequestValidationError is the validation error returned by
// NamedEntityListRequest.Validate if the designated constraints aren't met.
type NamedEntityListRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NamedEntityListRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NamedEntityListRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NamedEntityListRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NamedEntityListRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NamedEntityListRequestValidationError) ErrorName() string {
	return "NamedEntityListRequestValidationError"
}

// Error satisfies the builtin error interface
func (e NamedEntityListRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNamedEntityListRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NamedEntityListRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NamedEntityListRequestValidationError{}

// Validate checks the field values on NamedEntityIdentifierList with the rules
// defined in the proto definition for this message. If any rules are
// violated, an error is returned.
func (m *NamedEntityIdentifierList) Validate() error {
	if m == nil {
		return nil
	}

	for idx, item := range m.GetEntities() {
		_, _ = idx, item

		if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return NamedEntityIdentifierListValidationError{
					field:  fmt.Sprintf("Entities[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	// no validation rules for Token

	return nil
}

// NamedEntityIdentifierListValidationError is the validation error returned by
// NamedEntityIdentifierList.Validate if the designated constraints aren't met.
type NamedEntityIdentifierListValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NamedEntityIdentifierListValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NamedEntityIdentifierListValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NamedEntityIdentifierListValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NamedEntityIdentifierListValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NamedEntityIdentifierListValidationError) ErrorName() string {
	return "NamedEntityIdentifierListValidationError"
}

// Error satisfies the builtin error interface
func (e NamedEntityIdentifierListValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNamedEntityIdentifierList.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NamedEntityIdentifierListValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NamedEntityIdentifierListValidationError{}

// Validate checks the field values on NamedEntityList with the rules defined
// in the proto definition for this message. If any rules are violated, an
// error is returned.
func (m *NamedEntityList) Validate() error {
	if m == nil {
		return nil
	}

	for idx, item := range m.GetEntities() {
		_, _ = idx, item

		if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return NamedEntityListValidationError{
					field:  fmt.Sprintf("Entities[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	// no validation rules for Token

	return nil
}

// NamedEntityListValidationError is the validation error returned by
// NamedEntityList.Validate if the designated constraints aren't met.
type NamedEntityListValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NamedEntityListValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NamedEntityListValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NamedEntityListValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NamedEntityListValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NamedEntityListValidationError) ErrorName() string { return "NamedEntityListValidationError" }

// Error satisfies the builtin error interface
func (e NamedEntityListValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNamedEntityList.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NamedEntityListValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NamedEntityListValidationError{}

// Validate checks the field values on NamedEntityGetRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, an error is returned.
func (m *NamedEntityGetRequest) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for ResourceType

	if v, ok := interface{}(m.GetId()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return NamedEntityGetRequestValidationError{
				field:  "Id",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	return nil
}

// NamedEntityGetRequestValidationError is the validation error returned by
// NamedEntityGetRequest.Validate if the designated constraints aren't met.
type NamedEntityGetRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NamedEntityGetRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NamedEntityGetRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NamedEntityGetRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NamedEntityGetRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NamedEntityGetRequestValidationError) ErrorName() string {
	return "NamedEntityGetRequestValidationError"
}

// Error satisfies the builtin error interface
func (e NamedEntityGetRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNamedEntityGetRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NamedEntityGetRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NamedEntityGetRequestValidationError{}

// Validate checks the field values on NamedEntityUpdateRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, an error is returned.
func (m *NamedEntityUpdateRequest) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for ResourceType

	if v, ok := interface{}(m.GetId()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return NamedEntityUpdateRequestValidationError{
				field:  "Id",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if v, ok := interface{}(m.GetMetadata()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return NamedEntityUpdateRequestValidationError{
				field:  "Metadata",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	return nil
}

// NamedEntityUpdateRequestValidationError is the validation error returned by
// NamedEntityUpdateRequest.Validate if the designated constraints aren't met.
type NamedEntityUpdateRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NamedEntityUpdateRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NamedEntityUpdateRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NamedEntityUpdateRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NamedEntityUpdateRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NamedEntityUpdateRequestValidationError) ErrorName() string {
	return "NamedEntityUpdateRequestValidationError"
}

// Error satisfies the builtin error interface
func (e NamedEntityUpdateRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNamedEntityUpdateRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NamedEntityUpdateRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NamedEntityUpdateRequestValidationError{}

// Validate checks the field values on NamedEntityUpdateResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, an error is returned.
func (m *NamedEntityUpdateResponse) Validate() error {
	if m == nil {
		return nil
	}

	return nil
}

// NamedEntityUpdateResponseValidationError is the validation error returned by
// NamedEntityUpdateResponse.Validate if the designated constraints aren't met.
type NamedEntityUpdateResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NamedEntityUpdateResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NamedEntityUpdateResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NamedEntityUpdateResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NamedEntityUpdateResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NamedEntityUpdateResponseValidationError) ErrorName() string {
	return "NamedEntityUpdateResponseValidationError"
}

// Error satisfies the builtin error interface
func (e NamedEntityUpdateResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNamedEntityUpdateResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NamedEntityUpdateResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NamedEntityUpdateResponseValidationError{}

// Validate checks the field values on ObjectGetRequest with the rules defined
// in the proto definition for this message. If any rules are violated, an
// error is returned.
func (m *ObjectGetRequest) Validate() error {
	if m == nil {
		return nil
	}

	if v, ok := interface{}(m.GetId()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ObjectGetRequestValidationError{
				field:  "Id",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	return nil
}

// ObjectGetRequestValidationError is the validation error returned by
// ObjectGetRequest.Validate if the designated constraints aren't met.
type ObjectGetRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ObjectGetRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ObjectGetRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ObjectGetRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ObjectGetRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ObjectGetRequestValidationError) ErrorName() string { return "ObjectGetRequestValidationError" }

// Error satisfies the builtin error interface
func (e ObjectGetRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sObjectGetRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ObjectGetRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ObjectGetRequestValidationError{}

// Validate checks the field values on ResourceListRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, an error is returned.
func (m *ResourceListRequest) Validate() error {
	if m == nil {
		return nil
	}

	if v, ok := interface{}(m.GetId()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ResourceListRequestValidationError{
				field:  "Id",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for Limit

	// no validation rules for Token

	// no validation rules for Filters

	if v, ok := interface{}(m.GetSortBy()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ResourceListRequestValidationError{
				field:  "SortBy",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	return nil
}

// ResourceListRequestValidationError is the validation error returned by
// ResourceListRequest.Validate if the designated constraints aren't met.
type ResourceListRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ResourceListRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ResourceListRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ResourceListRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ResourceListRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ResourceListRequestValidationError) ErrorName() string {
	return "ResourceListRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ResourceListRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sResourceListRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ResourceListRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ResourceListRequestValidationError{}

// Validate checks the field values on EmailNotification with the rules defined
// in the proto definition for this message. If any rules are violated, an
// error is returned.
func (m *EmailNotification) Validate() error {
	if m == nil {
		return nil
	}

	return nil
}

// EmailNotificationValidationError is the validation error returned by
// EmailNotification.Validate if the designated constraints aren't met.
type EmailNotificationValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e EmailNotificationValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e EmailNotificationValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e EmailNotificationValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e EmailNotificationValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e EmailNotificationValidationError) ErrorName() string {
	return "EmailNotificationValidationError"
}

// Error satisfies the builtin error interface
func (e EmailNotificationValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sEmailNotification.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = EmailNotificationValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = EmailNotificationValidationError{}

// Validate checks the field values on PagerDutyNotification with the rules
// defined in the proto definition for this message. If any rules are
// violated, an error is returned.
func (m *PagerDutyNotification) Validate() error {
	if m == nil {
		return nil
	}

	return nil
}

// PagerDutyNotificationValidationError is the validation error returned by
// PagerDutyNotification.Validate if the designated constraints aren't met.
type PagerDutyNotificationValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e PagerDutyNotificationValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e PagerDutyNotificationValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e PagerDutyNotificationValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e PagerDutyNotificationValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e PagerDutyNotificationValidationError) ErrorName() string {
	return "PagerDutyNotificationValidationError"
}

// Error satisfies the builtin error interface
func (e PagerDutyNotificationValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sPagerDutyNotification.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = PagerDutyNotificationValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = PagerDutyNotificationValidationError{}

// Validate checks the field values on SlackNotification with the rules defined
// in the proto definition for this message. If any rules are violated, an
// error is returned.
func (m *SlackNotification) Validate() error {
	if m == nil {
		return nil
	}

	return nil
}

// SlackNotificationValidationError is the validation error returned by
// SlackNotification.Validate if the designated constraints aren't met.
type SlackNotificationValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SlackNotificationValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SlackNotificationValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SlackNotificationValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SlackNotificationValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SlackNotificationValidationError) ErrorName() string {
	return "SlackNotificationValidationError"
}

// Error satisfies the builtin error interface
func (e SlackNotificationValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSlackNotification.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SlackNotificationValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SlackNotificationValidationError{}

// Validate checks the field values on Notification with the rules defined in
// the proto definition for this message. If any rules are violated, an error
// is returned.
func (m *Notification) Validate() error {
	if m == nil {
		return nil
	}

	switch m.Type.(type) {

	case *Notification_Email:

		if v, ok := interface{}(m.GetEmail()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return NotificationValidationError{
					field:  "Email",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	case *Notification_PagerDuty:

		if v, ok := interface{}(m.GetPagerDuty()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return NotificationValidationError{
					field:  "PagerDuty",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	case *Notification_Slack:

		if v, ok := interface{}(m.GetSlack()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return NotificationValidationError{
					field:  "Slack",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	return nil
}

// NotificationValidationError is the validation error returned by
// Notification.Validate if the designated constraints aren't met.
type NotificationValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NotificationValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NotificationValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NotificationValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NotificationValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NotificationValidationError) ErrorName() string { return "NotificationValidationError" }

// Error satisfies the builtin error interface
func (e NotificationValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNotification.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NotificationValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NotificationValidationError{}

// Validate checks the field values on UrlBlob with the rules defined in the
// proto definition for this message. If any rules are violated, an error is returned.
func (m *UrlBlob) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for Url

	// no validation rules for Bytes

	return nil
}

// UrlBlobValidationError is the validation error returned by UrlBlob.Validate
// if the designated constraints aren't met.
type UrlBlobValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UrlBlobValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UrlBlobValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UrlBlobValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UrlBlobValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UrlBlobValidationError) ErrorName() string { return "UrlBlobValidationError" }

// Error satisfies the builtin error interface
func (e UrlBlobValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUrlBlob.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UrlBlobValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UrlBlobValidationError{}

// Validate checks the field values on Labels with the rules defined in the
// proto definition for this message. If any rules are violated, an error is returned.
func (m *Labels) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for Values

	return nil
}

// LabelsValidationError is the validation error returned by Labels.Validate if
// the designated constraints aren't met.
type LabelsValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e LabelsValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e LabelsValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e LabelsValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e LabelsValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e LabelsValidationError) ErrorName() string { return "LabelsValidationError" }

// Error satisfies the builtin error interface
func (e LabelsValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sLabels.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = LabelsValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = LabelsValidationError{}

// Validate checks the field values on Annotations with the rules defined in
// the proto definition for this message. If any rules are violated, an error
// is returned.
func (m *Annotations) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for Values

	return nil
}

// AnnotationsValidationError is the validation error returned by
// Annotations.Validate if the designated constraints aren't met.
type AnnotationsValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e AnnotationsValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e AnnotationsValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e AnnotationsValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e AnnotationsValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e AnnotationsValidationError) ErrorName() string { return "AnnotationsValidationError" }

// Error satisfies the builtin error interface
func (e AnnotationsValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sAnnotations.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = AnnotationsValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = AnnotationsValidationError{}

// Validate checks the field values on AuthRole with the rules defined in the
// proto definition for this message. If any rules are violated, an error is returned.
func (m *AuthRole) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for AssumableIamRole

	// no validation rules for KubernetesServiceAccount

	return nil
}

// AuthRoleValidationError is the validation error returned by
// AuthRole.Validate if the designated constraints aren't met.
type AuthRoleValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e AuthRoleValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e AuthRoleValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e AuthRoleValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e AuthRoleValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e AuthRoleValidationError) ErrorName() string { return "AuthRoleValidationError" }

// Error satisfies the builtin error interface
func (e AuthRoleValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sAuthRole.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = AuthRoleValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = AuthRoleValidationError{}

// Validate checks the field values on RawOutputDataConfig with the rules
// defined in the proto definition for this message. If any rules are
// violated, an error is returned.
func (m *RawOutputDataConfig) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for OutputLocationPrefix

	return nil
}

// RawOutputDataConfigValidationError is the validation error returned by
// RawOutputDataConfig.Validate if the designated constraints aren't met.
type RawOutputDataConfigValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e RawOutputDataConfigValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e RawOutputDataConfigValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e RawOutputDataConfigValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e RawOutputDataConfigValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e RawOutputDataConfigValidationError) ErrorName() string {
	return "RawOutputDataConfigValidationError"
}

// Error satisfies the builtin error interface
func (e RawOutputDataConfigValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sRawOutputDataConfig.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = RawOutputDataConfigValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = RawOutputDataConfigValidationError{}
